src/source/canvas_source.js

```diff
@@ -114,21 +115,29 @@ class CanvasSource extends ImageSource {
      */
     // setCoordinates inherited from ImageSource
 
-    readCanvas() {
+    readCanvas(resize: boolean) {
         // We *should* be able to use a pure HTMLCanvasElement in
         // texImage2D/texSubImage2D (in ImageSource#_prepareImage), but for
         // some reason this breaks the map on certain GPUs (see #4262).
 
         if (this.context instanceof CanvasRenderingContext2D) {
-            return this.context.getImageData(0, 0, this.width, this.height);
+            this.canvasData = this.context.getImageData(0, 0, this.width, this.height);
         } else if (this.context instanceof WebGLRenderingContext) {
             const gl = this.context;
             const data = new Uint8Array(this.width * this.height * 4);
             gl.readPixels(0, 0, this.width, this.height, gl.RGBA, gl.UNSIGNED_BYTE, data);
+
             if (!this.secondaryContext) this.secondaryContext = window.document.createElement('canvas').getContext('2d');
-            const imageData = this.secondaryContext.createImageData(this.width, this.height);
-            imageData.data.set(data);
-            return imageData;
+            if (!this.canvasData || resize) {
+                this.canvasData = this.secondaryContext.createImageData(this.width, this.height);
+            }
+
+            // WebGL reads pixels bottom to top, but for our ImageData object we need top to bottom: flip here
+            const flipped = new Uint8Array(this.width * this.height * 4);
+            for (let i = this.height - 1, j = 0; i >= 0; i--, j++) {
+                flipped.set(data.subarray(i * this.width * 4, (i + 1) * this.width * 4), j * this.width * 4);
+            }
+            this.canvasData.data.set(flipped);
```

Comment:
One more thing — is it possible to do `this.canvasData.data.set` inside the loop instead of using a temporary `flipped` array?